playground/alex_bot.py

```diff
@@ -0,0 +1,205 @@
+
+
+@dataclass
+class Window:
+    pattern: List[int]
+    position: List[Tuple[int, int]]
+
+
+class AlexBot:
+
+    def __init__(self, k: int = 6, m: int = 19, n: int = 19):
+        self.k = k
+        self.m = m
+        self.n = n
+
+    def put_stone(self, log: List[Stone]) -> Stone:
+        turn = "b" if (len(log) + 1) % 4 in [0, 1] else "w"
+        if len(log) == 0:
+            return Stone(str(10), str(10), turn)
+
+        if len(log) == 1:
+            next_x = int(log[0].x) + choice([-1, 1])
+            next_y = int(log[0].y) + choice([-1, 1])
+            return Stone(str(next_x), str(next_y), turn)
+
+        log_x = [int(stone.x) for stone in log]
+        log_y = [int(stone.y) for stone in log]
+
+        left_end = max(min(log_x) - self.k, 1)
+        right_end = min(max(log_x) + self.k, self.m)
+        upper_end = max(min(log_y) - self.k, 1)
+        lower_end = min(max(log_y) + self.k, self.n)
+
+        width = right_end - left_end + 1
+        height = lower_end - upper_end + 1
+
+        focus_area = np.zeros((height, width), dtype=int)
+        for stone in log:
+            normalize_row_index = int(stone.y) - upper_end
+            normalize_column_index = int(stone.x) - left_end
+            focus_area[normalize_row_index, normalize_column_index] = 1 if stone.color == "b" else -1
+
+        horizontal = self.find_connection(focus_area, "h", left_end, upper_end)
+        vertical = self.find_connection(focus_area, "v", left_end, upper_end)
+        negative_diagonal = self.find_connection(focus_area, "nd", left_end, upper_end)
+        positive_diagonal = self.find_connection(focus_area, "pd", left_end, upper_end)
+        candidates = horizontal + vertical + negative_diagonal + positive_diagonal
+        if len(candidates) == 0:
+            print("expect tie")
+            random_empty_spot = choice(self.empty_spots(log))
+            return Stone(random_empty_spot.x, random_empty_spot.y, turn)
+
+        max_stone = max([self.stone_sum(candidate) for candidate in candidates])
+        if max_stone >= self.k:
+            return Stone("1", "1", turn)
+        max_stone_candidates = [candidate for candidate in candidates if self.stone_sum(candidate) == max_stone]
+        max_stone_candidate = choice(max_stone_candidates)
+        window_index = (
+            sorted(
+                [i for i, x in enumerate(max_stone_candidate.pattern) if x == 0],
+                key=lambda x: abs(x - self.k / 2),
+            )[0]
+        )
+        position = max_stone_candidate.position[window_index]
+
+        return Stone(str(position[0]), str(position[1]), turn)
+
+    def find_connection(self,
+                        focus_area: np.ndarray,
+                        direction: str,
+                        left_end: int,
+                        upper_end: int,
+                        ) -> List[Window]:
+        connections = []
+        max_stone = 0
+
+        if direction == "h" or direction == "horizontal":
```

Comment:
directions might be widely used in other places, 
so maybe we can separate as constants and use them: https://www.edureka.co/community/16286/create-a-constant-in-python

Reply:
Then these direction constants and above `turn_check` function could be the items of our new common function code.